agent/php_user_instrument.c

```diff
@@ -88,10 +93,69 @@ int nr_zend_call_orig_execute_special(nruserfn_t* wraprec,
+ * instrumentation they want to monitor. `user_function_wrappers` is a vector of
+ * pointers to wrappers, after the zend function represented by a wraprec has
+ * the reserved field modified, the pointer to the wraprec (which again, exists
+ * in `nr_php_op_array_set_wraprec`) goes into the vector.
+ * `nr_php_op_array_set_wraprec` is always a superset of
+ * `user_function_wrappers` and the wraprec pointers that exist in
+ * `user_function_wrappers` always exist in `nr_php_op_array_set_wraprec`.
+ *
+ * `nr_php_op_array_set_wraprec` is populated a few different ways.
+ * 1)from function `nr_php_add_transaction_naming_function` called from
+ * `php_nrini.c` to set the naming for all the transactions the user  set in the
+ * ini with `newrelic.webtransaction.name.functions` 2) from function
+ * `nr_php_add_custom_tracer` from `php_nrini.c` to set the naming for all the
+ * transactions the user set in the ini with
+ * `newrelic.transaction_tracer.custom` 3)
+ * nr_php_user_function_add_declared_callback (prior to PHP 7.3) 4) from
+ * function `nr_php_wrap_user_function` called from php_wrapper sets the wraprec
+ * with framework specific instrumentation. 5) from function
+ * `nr_php_wrap_callable` (in `php_wrapper.c`) used only by Wordpress and predis
+ * for custom instrumentation that adds `is_transient` wrappers that get cleaned
+ * up with each shutdown.
+ *
+ * When overwriting the zend_execute_ex function, every effort was made to
+ * reduce performance overhead because until the agent returns control, we are
+ * the bottleneck of PHP execution on a customer's machine.  Overwriting the
+ * reserved field was seen as a quick way to check if a function is instrumented
+ * or not.
+ *
+ * However, with PHP 8+, we've begun noticing more conflicts with the reserved
+ * fields.  Additionally, as we are no longer halting execution while we
+ * process, we can search through `nr_php_op_array_set_wraprec` instead of
+ * setting reserved field and getting from the `user_function_wrappers` vector.
+ * This stops the issues (segfaults, incorrect naming in laravel, etc) that we
+ * were observing, especially with PHP 8.1.
  */
 static void nr_php_wrap_zend_function(zend_function* func,
                                       nruserfn_t* wraprec TSRMLS_DC) {
+#if ZEND_MODULE_API_NO >= ZEND_7_4_X_API_NO
+  const char* filename = nr_php_function_filename(func);
+  /*
+   * Before setting a filename, ensure it is not NULL and it doesn't equal the
+   * "special" designation given to funcs without filenames. If the function is
+   * an evaluated expression or called directly from the CLI there is no
+   * filename, but the function says the filename is "-". Avoid setting in this
+   * case; otherwise, all the evaluated/cli calls would match.
+   */
+  if ((NULL == wraprec->filename) && (NULL != filename)
+      && (0 != nr_strcmp("-", filename))) {
+    wraprec->filename = nr_strdup(filename);
+  }
+
+  wraprec->lineno = nr_php_zend_function_lineno(func);
```

Comment:
What is we move this as the first check?
If it is zero, I believe we can skip the filename lookup/strdups.

Reply:
You mean that if filename is '-' then lineno is 0? Is reverse true: if lineno is 0 will the filename be '-' or can it be something different, useful, meaningful?

Reply:
This function is infrequently called and the strcomp is trivial because it is only 1 character.